File: trunk/proj/CoreLibrary/src/com/jasonzqshen/familyaccounting/core/transaction/GLAccountBalanceCollection.java
package com.jasonzqshen.familyaccounting.core.transaction;

import java.util.Hashtable;

import com.jasonzqshen.familyaccounting.core.CoreDriver;
import com.jasonzqshen.familyaccounting.core.listeners.CreateMasterDataListener;
import com.jasonzqshen.familyaccounting.core.listeners.LoadDocumentListener;
import com.jasonzqshen.familyaccounting.core.listeners.LoadMasterDataListener;
import com.jasonzqshen.familyaccounting.core.listeners.SaveDocumentListener;
import com.jasonzqshen.familyaccounting.core.masterdata.GLAccountMasterData;
import com.jasonzqshen.familyaccounting.core.masterdata.MasterDataBase;
import com.jasonzqshen.familyaccounting.core.masterdata.MasterDataFactoryBase;
import com.jasonzqshen.familyaccounting.core.masterdata.MasterDataIdentity_GLAccount;
import com.jasonzqshen.familyaccounting.core.masterdata.MasterDataManagement;
import com.jasonzqshen.familyaccounting.core.masterdata.MasterDataType;
import com.jasonzqshen.familyaccounting.core.utils.CreditDebitIndicator;
import com.jasonzqshen.familyaccounting.core.utils.CurrencyAmount;
import com.jasonzqshen.familyaccounting.core.utils.GLAccountGroup;

public class GLAccountBalanceCollection {
	private final CoreDriver _coreDriver;

	private final Hashtable<MasterDataIdentity_GLAccount, GLAccountBalanceItem> _items;

	GLAccountBalanceCollection(CoreDriver coreDriver) {
		_coreDriver = coreDriver;

		_items = new Hashtable<MasterDataIdentity_GLAccount, GLAccountBalanceItem>();

		_coreDriver.getListenersManagement().addSaveDocListener(
				_saveDocumentListener);
		_coreDriver.getListenersManagement().addLoadDocListener(
				_loadDocumentListener);
		_coreDriver.getListenersManagement().addLoadMasterListener(
				_loadMasterDataListener);
		coreDriver.getListenersManagement().addCreateMasterListener(
				_createMasterListener);
	}

	// load document listener
	private LoadDocumentListener _loadDocumentListener = new LoadDocumentListener() {
		public void onLoadDocumentListener(Object source, HeadEntity document) {
			newDoc(document);
		}
	};

	// save document listener
	private SaveDocumentListener _saveDocumentListener = new SaveDocumentListener() {
		public void onSaveDocumentListener(HeadEntity document) {
			newDoc(document);
		}
	};

	private LoadMasterDataListener _loadMasterDataListener = new LoadMasterDataListener() {
		public void onLoadMasterDataListener(Object obj,
				MasterDataBase masterData) {
			newMasterdata(masterData);
		}
	};

	private CreateMasterDataListener _createMasterListener = new CreateMasterDataListener() {
		public void onCreateMasterDataListener(MasterDataFactoryBase factory,
				MasterDataBase master) {
			newMasterdata(master);
		}

	};

	/**
	 * set report when new document
	 * 
	 * @param head
	 */
	private void newDoc(HeadEntity head) {

		for (ItemEntity item : head.getItems()) {
			CurrencyAmount amount = item.getAmount();
			if (item.getCDIndicator() == CreditDebitIndicator.CREDIT) {
				amount.negate();
			}
			GLAccountBalanceItem balItem = _items.get(item.getGLAccount());

			balItem.addAmount(head.getMonthId(), amount);
		}
	}

	/**
	 * set report when new master data
	 * 
	 * @param dataS
	 */
	private void newMasterdata(MasterDataBase data) {
		if (data instanceof GLAccountMasterData) {
			MasterDataIdentity_GLAccount glId = (MasterDataIdentity_GLAccount) data
					.getIdentity();
			_items.put(glId, new GLAccountBalanceItem(glId));
		}
	}

	/**
	 * get balance item
	 */
	public GLAccountBalanceItem getBalanceItem(
			MasterDataIdentity_GLAccount glAccount) {
		return _items.get(glAccount);
	}

	/**
	 * get group balance
	 * 
	 * @param accountGroup
	 * @return
	 */
	public CurrencyAmount getGroupBalance(GLAccountGroup accountGroup) {
		CurrencyAmount ret = new CurrencyAmount();
		MasterDataManagement mdMgmt = _coreDriver.getMasterDataManagement();
		for (MasterDataIdentity_GLAccount id : _items.keySet()) {
			GLAccountMasterData glAccount = (GLAccountMasterData) mdMgmt
					.getMasterData(id, MasterDataType.GL_ACCOUNT);
			if (glAccount.getGroup() == accountGroup) {
				ret.addTo(_items.get(id).getSumAmount());
			}
		}

		return ret;
	}

	/**
	 * get group balance
	 * 
	 * @param accountGroup
	 * @return
	 */
	public CurrencyAmount getGroupBalance(GLAccountGroup accountGroup,
			MonthIdentity startMonthId, MonthIdentity endMonthId) {
		CurrencyAmount ret = new CurrencyAmount();
		MasterDataManagement mdMgmt = _coreDriver.getMasterDataManagement();
		for (MasterDataIdentity_GLAccount id : _items.keySet()) {
			GLAccountMasterData glAccount = (GLAccountMasterData) mdMgmt
					.getMasterData(id, MasterDataType.GL_ACCOUNT);
			if (glAccount.getGroup() == accountGroup) {
				ret.addTo(_items.get(id).getSumAmount(startMonthId, endMonthId));
			}
		}

		return ret;
	}
}

File: proj/WinStore_FamilyAccounting/WinStore_FamilyAccountingCore/Transaction/GLAccountBalanceCollection.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using WinStore_FamilyAccountingCore.Exceptions;
using WinStore_FamilyAccountingCore.MasterData;
using WinStore_FamilyAccountingCore.Utilities;

namespace WinStore_FamilyAccountingCore.Transaction
{
    public class GLAccountBalanceCollection
    {
        private readonly CoreDriver _coreDriver;

        private readonly MasterDataManagement _mdMgmt;
        public MasterDataManagement MdMgmt { get { return _mdMgmt; } }

        private readonly Dictionary<MasterDataIdentity_GLAccount, GLAccountBalanceItem> _items;

        internal GLAccountBalanceCollection(CoreDriver coreDriver,
                MasterDataManagement mdMgmt)
        {
            _coreDriver = coreDriver;
            _mdMgmt = mdMgmt;

            _items = new Dictionary<MasterDataIdentity_GLAccount, GLAccountBalanceItem>();

            _coreDriver.ListenerMgmt.DocumentSavedHandler += documentSaved;
            _coreDriver.ListenerMgmt.DocumentLoadHandler += documentLoad;
            _coreDriver.ListenerMgmt.MasterDataLoadHandler += masterDataLoad;
            _coreDriver.ListenerMgmt.MasterDataCreatedHandler += masterDataCreated;
            _coreDriver.ListenerMgmt.DocumentReversedHandler += documentReversed;
            _coreDriver.ListenerMgmt.GLAccountInitAmountChangedHandler += glAccountInitAmountChanged;
        }

        #region Events handler
        private void documentLoad(Object source, HeadEntity document)
        {
            newDoc(document);
        }

        private void documentReversed(HeadEntity doc)
        {
            reverseDoc(doc);
        }

        private void documentSaved(HeadEntity document)
        {
            newDoc(document);
        }
        private void masterDataLoad(Object source, MasterDataBase masterData)
        {
            newMasterdata(masterData);
        }

        private void masterDataCreated(MasterDataFactoryBase factory, MasterDataBase master)
        {
            newMasterdata(master);
        }

        private void glAccountInitAmountChanged(MasterDataIdentity_GLAccount glAccountId
            , CurrencyAmount orgAmount, CurrencyAmount newAmount)
        {
            GLAccountBalanceItem item;
            if (!_items.TryGetValue(glAccountId, out item))
            {
                return;
            }

            // ret += newAmount - orgAmount
            item.AddAmount(_coreDriver.StartMonthId
                , CurrencyAmount.Minus(newAmount, orgAmount));
        }
        #endregion


        /// <summary>
        /// reverse document
        /// </summary>
        /// <param name="doc"></param>
        private void reverseDoc(HeadEntity doc)
        {
            foreach (ItemEntity item in doc.DocItems)
            {
                CurrencyAmount amount = item.Amount;
                if (item.CdIndicator == CreditDebitIndicator.DEBIT)
                {
                    amount.Negate();
                }
                GLAccountBalanceItem balItem;
                if (!_items.TryGetValue(item.GLAccount, out balItem))
                {
                    continue;
                }

                balItem.AddAmount(doc.MonthID, amount);
            }
        }
        /// <summary>
        /// set report when new doc
        /// </summary>
        /// <param name="head"></param>
        private void newDoc(HeadEntity head)
        {
            foreach (ItemEntity item in head.DocItems)
            {
                CurrencyAmount amount = item.Amount;
                if (item.CdIndicator == CreditDebitIndicator.CREDIT)
                {
                    amount.Negate();
                }
                GLAccountBalanceItem balItem;
                if (!_items.TryGetValue(item.GLAccount, out balItem))
                {
                    continue;
                }

                balItem.AddAmount(head.MonthID, amount);
            }
        }

        /// <summary>
        /// set report when new master data
        /// </summary>
        /// <param name="data"></param>
        private void newMasterdata(MasterDataBase data)
        {
            GLAccountMasterData glAcc = data as GLAccountMasterData;
            if (glAcc == null)
            {
                return;
            }
            if (!_items.ContainsKey(glAcc.GLIdentity))
            {
                GLAccountBalanceItem item = new GLAccountBalanceItem(glAcc.GLIdentity);
                _items.Add(glAcc.GLIdentity, item);
                item.AddAmount(_coreDriver.StartMonthId, glAcc.InitAmount);
            }
        }

        /// <summary>
        /// get balance item
        /// </summary>
        /// <param name="glAccount"></param>
        /// <returns></returns>
        public GLAccountBalanceItem GetBalanceItem(
                MasterDataIdentity_GLAccount glAccount)
        {
            GLAccountBalanceItem item;
            if (!_items.TryGetValue(glAccount, out item))
            {
                return null;
            }
            return item;
        }

        /// <summary>
        /// Get group balance amount
        /// </summary>
        /// <param name="accountGroup"></param>
        /// <returns></returns>
        /// <exception cref="SystemException"></exception>
        public CurrencyAmount GetGroupBalance(GLAccountGroupENUM accountGroup)
        {
            CurrencyAmount ret = new CurrencyAmount();
            MasterDataManagement mdMgmt = this._mdMgmt;
            foreach (var item in _items)
            {
                GLAccountMasterData glAccount;
                try
                {
                    glAccount = (GLAccountMasterData)mdMgmt
                           .GetMasterData(item.Key, MasterDataType.GL_ACCOUNT);
                }
                catch (Exception e)
                {
                    _coreDriver.logDebugInfo(this.GetType(), 183, e.Message, MessageType.ERRO);
                    throw new SystemException(e);
                }
                if (glAccount.Group.Identity == accountGroup)
                {
                    ret.AddTo(item.Value.Sum);
                }
            }

            return ret;
        }

        /// <summary>
        /// Get group balance from start month to end month
        /// </summary>
        /// <param name="accountGroup"></param>
        /// <param name="startMonthId"></param>
        /// <param name="endMonthId"></param>
        /// <returns></returns>
        /// <exception cref="SystemException"></exception>
        public CurrencyAmount GetGroupBalance(GLAccountGroupENUM accountGroup,
            MonthIdentity startMonthId, MonthIdentity endMonthId)
        {
            CurrencyAmount ret = new CurrencyAmount();
            MasterDataManagement mdMgmt = this._mdMgmt;
            foreach (var item in _items)
            {
                GLAccountMasterData glAccount;
                try
                {
                    glAccount = (GLAccountMasterData)mdMgmt
                           .GetMasterData(item.Key, MasterDataType.GL_ACCOUNT);
                }
                catch (Exception e)
                {
                    _coreDriver.logDebugInfo(this.GetType(), 183, e.Message, MessageType.ERRO);
                    throw new SystemException(e);
                }
                if (glAccount.Group.Identity == accountGroup)
                {
                    ret.AddTo(item.Value.GetSumAmount(startMonthId, endMonthId));
                }
            }

            return ret;
        }

    }
}
